io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py

```diff
@@ -77,6 +79,7 @@ def export_tri_shapes(self, b_obj, n_parent, trishape_name=None):
 
         # get mesh from b_obj
         b_mesh = self.get_triangulated_mesh(b_obj)
+        b_mesh.calc_normals_split()
```

Comment:
I use calc_tangents() later, too, but only if the tangents need to be exported.

Reply:
I see. Not sure if blender avoids recalculating the loop normals data in this case.

Reply:
Not sure. I tried it on a simple subdivided cube with 49000 vertices (close to the nif limit of ~65000 vertices). Calc_normals was still too fast to measure (just displayed as 0.0 s) whereas calc_tangents() was always approximately 0.05 s. Given that calc_normals_split() is so fast, it's not really possible to check by timing whether calc_tangents does it again (but then it doesn't really matter whether it does). Based on the code of `rna_Mesh_calc_tangents()`, I'd wager it does avoid recalculating.

Either way, since calc_normals_split() seems to be so much faster, I reckon it'd be worth not calling calc_tangents() even if that did cause calc_normals_split() to be called twice.

Reply:
Seems like the times are pretty good, although I don't know how beefy your hardware is.
Happy enough for now as we are still in the "I want this feature", rather than "this feature and its talking to long to run" phase. I am sure there are plenty of bottleknecks which I suppose NumPy may end up solving in the long run.